wgpu-core/src/any_surface.rs

```diff
@@ -1,92 +1,93 @@
+    /// If `self` refers to an `A::Surface`, returns a reference to it.
+    pub fn downcast_ref<A: HalApi>(&self) -> Option<&A::Surface> {
+        if A::VARIANT != self.vtable.backend {
+            return None;
+        }
+
+        // SAFETY: We just checked the instance above implicitly by the backend
+        // that it was statically constructed through.
+        Some(unsafe { &*self.data.as_ptr().cast::<A::Surface>() })
     }
 
-    /// If `self` is an `Arc<HalSurface<A>>`, returns that.
-    pub fn take<A: HalApi>(self) -> Option<Arc<HalSurface<A>>> {
-        // `Arc::downcast` returns `Arc<T>`, but requires that `T` be `Sync` and
-        // `Send`, and this is not the case for `HalSurface` in wasm builds.
-        //
-        // But as far as I can see, `Arc::downcast` has no particular reason to
-        // require that `T` be `Sync` and `Send`; the steps used here are sound.
-        if (self.0).is::<HalSurface<A>>() {
-            // Turn the `Arc`, which is a pointer to an `ArcInner` struct, into
-            // a pointer to the `ArcInner`'s `data` field. Carry along the
-            // vtable from the original `Arc`.
-            let raw_erased: *const (dyn Any + 'static) = Arc::into_raw(self.0);
-            // Remove the vtable, and supply the concrete type of the `data`.
-            let raw_typed: *const HalSurface<A> = raw_erased.cast::<HalSurface<A>>();
-            // Convert the pointer to the `data` field back into a pointer to
-            // the `ArcInner`, and restore reference-counting behavior.
-            let arc_typed: Arc<HalSurface<A>> = unsafe {
-                // Safety:
-                // - We checked that the `dyn Any` was indeed a `HalSurface<A>` above.
-                // - We're calling `Arc::from_raw` on the same pointer returned
-                //   by `Arc::into_raw`, except that we stripped off the vtable
-                //   pointer.
-                // - The pointer must still be live, because we've borrowed `self`,
-                //   which holds another reference to it.
-                // - The format of a `ArcInner<dyn Any>` must be the same as
-                //   that of an `ArcInner<HalSurface<A>>`, or else `AnyHalSurface::new`
-                //   wouldn't be possible.
-                Arc::from_raw(raw_typed)
-            };
-            Some(arc_typed)
-        } else {
-            None
+    /// If `self` is an `Arc<A::Surface>`, returns that.
+    pub fn take<A: HalApi>(self) -> Option<A::Surface> {
+        if A::VARIANT != self.vtable.backend {
+            return None;
         }
+
+        // Disable drop glue, since we're returning the owned surface. The
+        // caller will be responsible for dropping it.
+        let this = ManuallyDrop::new(self);
+
+        // SAFETY: We just checked the instance above implicitly by the backend
+        // that it was statically constructed through.
+        Some(unsafe { *Box::from_raw(this.data.as_ptr().cast::<A::Surface>()) })
```

Comment:
Unfortunately this doesn't prevent the destructor from being called, the `data` field is copied out. Note that the drop implementation we want to prevent from running is `self`, not `data`.

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d87d79f3bd852e80c7db13e27d2d87a7).

Reply:
Oh, derp, you're totally right, and I forgot to consider that interaction with `Copy`. 😅 I'll `Resolve` this.

Non-blocking thoughts: I would suggest `std::mem::forget` after the _fully_ destructuring, but I don't consider that _significantly_ better than `ManuallyDrop` in this particular case. I would consider it for anything more complicated, since destructuring can help a programmer remember to handle other fields or otherwise have a signal that other fields might need to be handled.